### PR TITLE
Recognize target class within ->, ->>, doto

### PR DESCRIPTION
**Pull request by vemv**
> Closes https://github.com/alexander-yakushev/compliment/issues/61

Introduces support for ->, ->> and doto, by macroexpanding things in a sensible and constrained manner.

~(also .., but code coverage is pending atm)~

Other refinements are introduced so that ->/doto is actually useful. In specific terms, types are now better propagated/perceived when encountering vars, and `let` bindings (note: doto expands to a `let`, which is why it's important), etc.

Build is green:

image

Looking forward to feedback.

Cheers - V

### CODE REVIEW

**Review by alexander-yakushev — COMMENTED** [review]
Thanks for the PR. There is a lot to go through, won't manage in a single go. A few comments so far.

File: src/compliment/context.clj [hunk 1]
```diff
@@ -3,15 +3,36 @@
   (:require [clojure.string :as str]
             [clojure.walk :as walk]))
 
+(defn- walk-meta-preserving
```
[the remark this answers is not in the document]
Comment (alexander-yakushev) [reply src/compliment/context.clj:6]: The comment looked suspiciously familiar:).

File: test/compliment/sources/t_class_members.clj [hunk 2]
```diff
@@ -35,60 +83,61 @@
...
+    (strip-tags (src/members-candidates ".p" (-ns) (ctx/cache-context
+                                                    "(__prefix__ ^java.util.Map foo)")))
     => (just [".putAll" ".putIfAbsent" ".put"] :in-any-order)
 
-    (strip-tags (src/members-candidates ".in" (-ns) (ctx/parse-context
-                                                     '(let [^Thread foo ...
-                                                            ... ...
-                                                            _ (.start foo)]
-                                                        (__prefix__ foo)))))
+    (strip-tags (src/members-candidates ".in" (-ns) (ctx/cache-context
```
[the remark this answers is not in the document]
Comment (alexander-yakushev) [reply test/compliment/sources/t_class_members.clj:106]: Saw your other comment. Yeah, passing strings is more realistic since this is what clients do. But keeping forms in the tests is easier to manage (edit, indent).

I forgot about the problem with metadata in Fudje, gotta take a look.

File: src/compliment/sources/local_bindings.clj [hunk 3]
```diff
@@ -38,17 +66,26 @@
...
-        [binding-node]))
+        (let [result binding-node
+              candidate (delay (extract-tag-from-bound-to ns bound-to))]
+          (cond-> result
+            (and (not (-> result meta :tag))
+                 @candidate)
+            (vary-meta assoc :tag @candidate)
+
+            true
+            vector))))
```
Comment (alexander-yakushev) [inline src/compliment/sources/local_bindings.clj:77]: Pretty sure this can be rewritten without `delay` and `cond->`, will be clearer.

File: src/compliment/sources/local_bindings.clj [hunk 4]
```diff
@@ -1,30 +1,58 @@
...
-(def defn-like-forms '#{defn defn- fn defmacro defmethod})
-
-(def doseq-like-forms '#{doseq for})
-
-(def letfn-like-forms '#{letfn})
+            [compliment.sources.ns-mappings :refer [var-symbol? dash-matches?]]
+            [compliment.utils :as utils]))
+
+;; syntax-quoted variants are important for when we perform macroexpansion as part of our processing.
+(def let-like-forms #{'let `let
```
Comment (alexander-yakushev) [inline src/compliment/sources/local_bindings.clj:8]: Perhaps, instead of doing this, it makes sense to always drop the namespace from the symbol and only check for the symbol name. If there are collisions, let it be, may be even better.

I suggest you do it in a separate PR and we'll merge it before this.

File: src/compliment/utils.clj [hunk 5]
```diff
@@ -241,3 +243,32 @@ Note that should always have the same value, regardless of OS."
         ;; resource pathes always use "/" regardless of platform
         (.. (ensure-no-leading-slash file)
             (replace File/separator resource-separator))))))
+
+(defn var->class
+  "Given a form that may be a var, returns the class that is associated to its :tag or its value (in that precedence order)."
+  [ns form]
+  (when-let [var-ref (and (symbol? form)
+                          (when-let [found (some->> form (ns-resolve ns))]
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:251]: Not sure why you need `some->>` here.
Same for `when-let`, you check for `var?` later, it handles `nil` just fine.

File: src/compliment/utils.clj [hunk 6]
```diff
@@ -241,3 +243,32 @@ Note that should always have the same value, regardless of OS."
...
+        (class (deref var-ref)))))
+
+(defn invocation-form->class
+  "Given a form that might be an invocation form (i.e. a list),
+  return the class that is returned, according to the invoked function's var metadata."
+  [ns form]
+  (when-let [var-from-invocation (and (seq? form)
+                                      (symbol? (first form))
+                                      (ns-resolve ns (first form)))]
+    (and (= (class var-from-invocation) Var)
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:267]: and -> when, it should return `nil`, not `false`.

File: src/compliment/utils.clj [hunk 7]
```diff
@@ -241,3 +243,38 @@ Note that should always have the same value, regardless of OS."
...
+  (when-let [var-from-invocation (and (list-like? form)
+                                      (symbol? (first form))
+                                      (ns-resolve ns (first form)))]
+    (and (= (class var-from-invocation) Var)
+         (-> var-from-invocation meta :tag))))
+
+(defn java-interop->class
+  "Given a `form` that may represent a Java interop form (e.g. constructor call, method call), return its class."
+  [ns form]
+  ;; TODO implement
```
[the remark this answers is not in the document]
Comment (alexander-yakushev) [reply src/compliment/utils.clj:273]: I'd rather you remove it for now and add later when it's ready.

File: src/compliment/sources/local_bindings.clj [hunk 8]
```diff
@@ -68,13 +68,11 @@
         (not (#{'& '_} binding-node))
         (let [result binding-node
               candidate (delay (extract-tag-from-bound-to ns bound-to))]
```
Comment (alexander-yakushev) [inline src/compliment/sources/local_bindings.clj:79]: You left the delay here. Also, is there a reason to rename binding-node to result? Because it is shorter?

File: src/compliment/sources/local_bindings.clj [hunk 9]
```diff
@@ -68,13 +68,11 @@
...
-          (cond-> result
-            (and (not (-> result meta :tag))
-                 @candidate)
-            (vary-meta assoc :tag @candidate)
-
-            true
-            vector))))
+          (if (-> result meta :tag)
```
Comment (alexander-yakushev) [inline src/compliment/sources/local_bindings.clj:80]: Can just wrap the whole form in vector once.

File: src/compliment/sources/local_bindings.clj [hunk 10]
```diff
@@ -1,30 +1,67 @@
...
+                    "fn"
+                    "fn*"
+                    "defmacro"
+                    "defmethod"}
+                  (name x))))
+
+(defn- doseq-like-form? [x]
+  (and (symbol? x)
+       (contains? #{"doseq" "for"}
+                  (name x))))
```
Comment (alexander-yakushev) [inline src/compliment/sources/local_bindings.clj:33]: No need for newline here and in the next function.

File: src/compliment/utils.clj [hunk 11]
```diff
@@ -1,8 +1,10 @@
 (ns compliment.utils
   "Functions and utilities for source implementations."
-  (:import java.io.File java.nio.file.Files
-           (java.util.jar JarFile JarEntry)
-           java.util.function.Consumer))
+  (:import [clojure.lang Cons LazySeq Var]
+           [java.io File]
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:4]: I prefer `()` for imports, and spelling a FQCN directly if there is a single class imported from the package.

File: src/compliment/utils.clj [hunk 12]
```diff
@@ -241,3 +243,26 @@ Note that should always have the same value, regardless of OS."
         ;; resource pathes always use "/" regardless of platform
         (.. (ensure-no-leading-slash file)
             (replace File/separator resource-separator))))))
+
+(defn var->class
+  "Given a form that may be a var, returns the class that is associated to its :tag or its value (in that precedence order)."
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:248]: Docstring not 80-split.

File: src/compliment/utils.clj [hunk 13]
```diff
@@ -241,3 +243,26 @@ Note that should always have the same value, regardless of OS."
...
+  [ns form]
+  (when-let [var-ref (and (symbol? form)
+                          (let [found (ns-resolve ns form)]
+                            (when (var? found)
+                              found)))]
+    ;; let :tag take precedence - maybe the :tag says "this is an IFoo" (interface),
+    ;; and the class says "this is a FooImpl" (concrete class, which maybe has worse documentation
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:255]: This line pokes out too.

File: src/compliment/utils.clj [hunk 14]
```diff
@@ -241,3 +243,26 @@ Note that should always have the same value, regardless of OS."
...
+        (class (deref var-ref)))))
+
+(defn invocation-form->class
+  "Given a form that might be an invocation form (i.e. a list),
+  return the class that is returned, according to the invoked function's var metadata."
+  [ns form]
+  (when-let [var-from-invocation (and (seq? form)
+                                      (symbol? (first form))
+                                      (ns-resolve ns (first form)))]
+    (when (= (class var-from-invocation) Var)
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:267]: Why the code above uses `var?` but here this?

**Review by alexander-yakushev — COMMENTED** [review]
A few minor fixes and I think that would be all.

**Comment by vemv** [comment]
(jumping back soon)

**Comment by vemv** [comment]
All done! + Changelog

File: src/compliment/utils.clj [hunk 15]
```diff
@@ -1,8 +1,10 @@
 (ns compliment.utils
   "Functions and utilities for source implementations."
-  (:import java.io.File java.nio.file.Files
-           (java.util.jar JarFile JarEntry)
-           java.util.function.Consumer))
+  (:import (clojure.lang Cons LazySeq)
```
Comment (alexander-yakushev) [inline src/compliment/utils.clj:3]: This import is not used now.

**Comment by alexander-yakushev** [comment]
One last nit, and good to merge. Do you want to just squash+merge? Or do you plan to rebase, preserving multiple meaningful commits?

**Comment by vemv** [comment]
Cheers. This one is good to squash.

A new release will be happily included into cider-nrepl 🍏🍺

**Comment by alexander-yakushev** [comment]
Perfect, thanks! 🍻
